### ISSUE TITLE
Project Manager - Add capability to load using different Godot installs

### ISSUE THREAD

**Issue by awardell** [godotengine/godot-proposals#8370]
### Describe the project you are working on

Various projects created with various versions of Godot.

### Describe the problem or limitation you are having in your project

When I open Godot, sometimes I will see a little ⚠ icon for the project I intended to open. That means I clicked on the wrong version of the Godot Project Manager. I need to close it and open the right version, then open the project.

It would be convenient if I could select my Godot version I am going to use from within the Project Manager.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

See below.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

![image](https://github.com/godotengine/godot-proposals/assets/1361278/cf5d272a-6505-420f-91ef-969ff7a73ebc)

In the top-right corner there is already a display of the currently loaded Godot version. This could be changed into a drop-down menu to allow the user to select which version they want to launch using. When the user does so, the project list will update to add or remove the ⚠ icon as appropriate.

"How will the Project Manager know what Godot versions are installed on the machine?"

A list of install directories can be kept in the user data folder next to the list of project directories. When the Godot editor launches, it will add its own directory to that list if it is not already present.

The dropdown will display all Godot versions found in those directories. This way, if a user replaces the executable in a directory, the new executable will still be available. The Project Manager will extract what info it can from the executable filename. Maybe if it doesn't have "godot" somewhere in its name, it won't be displayed.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would probably be utilized by many. Certainly users can come up with their own ways of making launching Godot more convenient, but it would be a nice feature to have.

### Is there a reason why this should be core and not an add-on in the asset library?

Having it as an add-on would defeat the purpose if you opened a version which didn't have the add-on.

**Comment by DxUr** [comment]
I this will make your life more complicated, what about mistakenly open a project with a wrong version?

**Comment by Calinou** [comment]
> The dropdown will display all Godot versions found in those directories. This way, if a user replaces the executable in a directory, the new executable will still be available. The Project Manager will extract what info it can from the executable filename. Maybe if it doesn't have "godot" somewhere in its name, it won't be displayed.

Many people rename executables to something like just `godot.exe`, so this information should be gathered by calling the executable with `--version` and reading stdout.

**Comment by DxUr** [comment]
I think making another project manager that handle this is better like blender. For me it's a good idea but I don't have much time to work around. You can make one using any framework/language even using godot itself. It will be just a project manager.

**Comment by Calinou** [comment]
> I think making another project manager that handle this is better like blender. For me it's a good idea but I don't have much time to work around. You can make one using any framework/language even using godot itself. It will be just a project manager.

Indeed, this begs the question whether we should maintain an official version manager like [Hourglass](https://hourglass.jwestman.net/). It has many benefits, such as making the [Godot Flatpak](https://github.com/flathub/org.godotengine.Godot/) able to use official engine binaries and support using multiple Godot versions at once. The downside is that it requires maintaining a separate Godot project that acts as the Godot version, ensuring it's updated regularly, etc. It would add more burden to the release process.

**Comment by DxUr** [comment]
@Calinou This means making something like un**y hub. It's very beneficial as you will install just the project manager (aka Godot hub) then download any version you want. Especially in Debian and fedora because some times people can't find the latest version, even snapshots will be there this make it easy to manage things

**Comment by JoNax97** [comment]
There's this project:
https://github.com/MakovWait/godots
That basically has all the mentioned features and workflows. Plus the UX is very close to Godot itself, feels right at home.

Maybe some degree of official recognition would help spread adoption? (The name is a pain to Google, though)

**Comment by DxUr** [comment]
> There's this project: https://github.com/MakovWait/godots That basically has all the mentioned features and workflows. Plus the UX is very close to Godot itself, feels right at home.
> 
> Maybe some degree of official recognition would help spread adoption? (The name is a pain to Google, though)

It looks amazing so I think this branch should be closed!?

**Comment by awardell** [comment]
Well, godots certainly does implement this. I think I will begin using it, so thank you for the find, @JoNax97 

I would be happier with a project manager officially maintained by and packaged with godot, but godots certainly services. I have a couple of gripes with godots directly:

- It doesn't detect custom builds of godot when scanning. Not a huge deal.
- Binding editor version to a project is a manual process, and when you open a project it asks "Are you sure you want to?" without actually indicating whether or not there is a version mismatch. Kinda scary? I wish it would do something like Calinou's suggestion:

>Many people rename executables to something like just godot.exe, so this information should be gathered by calling the executable with --version and reading stdout.

and then give me confidence that I'm opening with the correct version.

- Seemingly no premade binaries for Windows or Linux. Again, not a huge deal.

I won't be heartbroken if this proposal is closed. I'll leave that up to any maintainer's discretion.

**Comment by JoNax97** [comment]
Not sure if I'm understanding this correctly:

> * Seemingly no premade binaries for Windows or Linux. Again, not a huge deal.

But I grabbed the windows executable from here: https://github.com/MakovWait/godots/releases/tag/v1.2.1.stable
The latest version includes an auto-updater so it should be a pretty hands-off process

**Comment by Loweredgames** [comment]
please...

**Comment by JoNax97** [comment]
Please don't comment on proposals just to bump them or show support. It sends email notifications to everyone involved. Use the thumbs up reaction instead.

**Comment by xahon** [comment]
Have anyone used https://github.com/Astral-Sheep/GodotHub? If it's ok, then it could be updated and included into the official site. What do maintainers think of it?

**Comment by AThousandShips** [comment]
I don't think we should list unofficial plugins or tools on the main page, but it could be added to [awesome-godot](https://github.com/godotengine/awesome-godot)

**Comment by xahon** [comment]
> I don't think we should list unofficial plugins or tools on the main page, but it could be added to [awesome-godot](https://github.com/godotengine/awesome-godot)

Yes, I mean that this tool can be used as a base for a fork for an official tool, so it would save some time by avoiding making boring basics (like creating a window, controls, etc..). If there are a licensing issues, then a separated ground-up tool could be made by the reference. Most major engines already have their own management tools: Unity - Unity Hub, UE - Epic Launcher. I attached the link above to show that there is already a tool, Godot needs something like that and the maintainers need to consider more focus on this task. I'm not a maintainer, but if I were, I'd start shortly with making a research and an implementation plan. The management tool (IMO) will draw more attention to the product you are making. I think awesome-godot is awesome but for non-basic things and I consider a "hub" as basic thing for nowadays game engines

**Comment by AThousandShips** [comment]
There's not really any significant demand for this feature and it doesn't align with how Godot works, so there's no need for maintainers to prioritize this

The fact that that tool hasn't been updated in over 2 years also indicates low demand

**Comment by JoNax97** [comment]
Are you sure there's no demand? Most Godot devs I know of have turned to using one of the unofficial managers.

**Comment by AThousandShips** [comment]
To be more specific: There's no established demand, see this proposal having no support, and the linked tool being abandoned for over 2 years, I don't know how popular any other tools are as I am not aware of them, but I've not seen many people demand this

But regardless we can't act on demand that isn't indicated, there could be massive demand, or tiny demand, but all data points to minimal demand

**Comment by LoweredgamesDev** [comment]
> I'm posting my request here since it's similar to this one:

### Problem or limitation

Hello everyone, I'm an indie developer and I noticed that Godot doesn't currently have a version control system integrated directly into the launcher. Other engines like Unity have a version control system that allows them to use that specific version. 

~~_I didn't see this request here, if it's duplicated please let me know so I can support the request._~~

_sorry for my bad English._

<img width="1910" height="997" alt="Image" src="https://github.com/user-attachments/assets/5fbcb93d-df86-4edd-abba-2e54808204fa" />

### Proposed improvement

My proposal is to add a UI to the launcher that lets you decide which version to download or use. I have something similar.

<img width="1920" height="888" alt="Image" src="https://github.com/user-attachments/assets/c2e4b024-b687-4152-9b69-ebb297190e93" />

This has huge advantages for development. For example, in my case, I can update the game whenever I want and I don't have to rush or worry about updating according to Godot's release cycle. This could also be useful for large teams or for developing large games.

There's already [an unofficial launcher that does this, Godot Launcher](https://godotlauncher.org/), you can review it as an example. 

It seems to me that this functionality can already be implemented in a simple way since I saw that the "tools" are already there. It requires a concept and understanding the UI structure, but otherwise it can be implemented in the next version. This way I can use the official Godot.

Thanks. :)

**Comment by LoweredgamesDev** [comment]
> There's not really any significant demand for this feature and it doesn't align with how Godot works, so there's no need for maintainers to prioritize this
> 
> The fact that that tool hasn't been updated in over 2 years also indicates low demand

However, I think that as a feature it is very useful for the quality of life of large projects and does not force you to always update to the latest version, breaking the project.

**Comment by vmenezio** [comment]
> [...] it doesn't align with how Godot works [...]

I am a bit curious about this assertion . What about a install version management system would conflict with how Godot works?

I ask because, since I currently contribute to several Godot projects, each built with a different version, and each sometimes updating versions without my explicit knowledge or involvement, my usual workflow is something like.

1. Decide to work on a project;
2. Go into my OS search bar and look up Godot, boot up the first version that comes up;
3. Scroll through the project list and likely realize the random version of Godot I opened isn't the one I needed;
4. Memorize the version number I actually need and close the Godot project manager;
5. Go back to the OS search bar and fire up the correct version of Godot.

Alternatively, I'll sometimes navigate to the path of the project manually, open its `project.godot` in a text editor, and check the version there, before actually opening Godot.

Both solutions seem awkward having previously used with Unity Hub to interface with similarly disparate engine installs.

Is either of the workflows above the expected way to use Godot? Or am I missing the intended path entirely?

**Comment by AThousandShips** [comment]
What I mean is that Godot isn't installed in most cases, and it's never installed as part of our distribution on non-mobile platforms, meaning there just doesn't exist a well defined way to *find* Godot versions, it's simply a self-contained executable that lives wherever you run it for most platforms, so this can't work without adding some way to do that, or having to manually register the different versions

**Comment by vmenezio** [comment]
Could it not work in an analogous way to how the Template Manager works? It provides a built-in way to to install templates from within the manager (equivalent to downloading new Godot versions via the Project Manager, such that they get automatically tracked), but the user is also able to point to a local TPZ (equivalent to manually adding a pre-existing Godot version, that gets tracked from that point on).

<img width="485" height="422" alt="Godot Template Manager dialog" src="https://github.com/user-attachments/assets/0c752ab0-8e8b-4e86-a3d5-839025e47c7e" />

**Comment by AThousandShips** [comment]
The templates are stored in a dedicated location, the executable is not, it would mean you have to put it in a single location which goes against how the engine is designed to work, and would make it hard to work with when building from source for example
